File: 7_graph/经典题/枚举边或顶点/enumerateTriangles.ts
/* eslint-disable no-inner-declarations */
// EnumeratingC3
// 遍历无向图中的所有三元环(a,b,c) a<b<c
//  O(E^1.5)
// https://kopricky.github.io/code/Graph/EnumeratingC3.html

/**
 * 遍历无向图中的所有三元环(a,b,c),`a<b<c`.
 * @param n 顶点数.
 * @param edges 边集.每条边为无向边'u<->v',范围为[0,n).边数<=2e5.
 * @param callback 回调函数.参数为三元环的三个顶点.
 * @complexity O(E^1.5).
 */
function enumerateTriangles(
  n: number,
  edges: [u: number, v: number][] | number[][],
  callback: (a: number, b: number, c: number) => void
): void {
  let edgeCount = 0
  const adjList: number[][] = Array(n)
  for (let i = 0; i < n; i++) adjList[i] = []
  edges.forEach(([u, v]) => {
    // 无向边定向
    if (u < v) adjList[u].push(v)
    else adjList[v].push(u)
    edgeCount++
  })

  const threshold = Math.sqrt(edgeCount / 4) | 0
  const memo: number[][] = Array(n)
  for (let i = 0; i < n; i++) memo[i] = []
  const visited = new Uint8Array(n)

  // processHighDegree
  for (let i = 0; i < n; i++) {
    const adjI = adjList[i]
    if (adjI.length <= threshold) continue
    for (let j = 0; j < adjI.length; j++) visited[adjI[j]] = 1
    for (let j = 0; j < adjI.length; j++) {
      const u = adjI[j]
      const adjU = adjList[u]
      for (let k = 0; k < adjU.length; k++) {
        const v = adjU[k]
        if (visited[v]) callback(i, u, v)
      }
    }
    for (let j = 0; j < adjI.length; j++) visited[adjI[j]] = 0
  }

  // processLowDegree
  for (let i = 0; i < n; i++) {
    const adjI = adjList[i]
    if (adjI.length > threshold) continue
    for (let j = 0; j < adjI.length; j++) {
      const u = adjI[j]
      for (let k = 0; k < adjI.length; k++) {
        const v = adjI[k]
        if (v > u) memo[u].push(i * n + v)
      }
    }
  }
  for (let i = 0; i < n; i++) {
    const adjI = adjList[i]
    for (let j = 0; j < adjI.length; j++) visited[adjI[j]] = 1
    for (let j = 0; j < memo[i].length; j++) {
      const hash = memo[i][j]
      const a = (hash / n) | 0
      const b = hash % n
      if (visited[b]) callback(a, i, b)
    }
    for (let j = 0; j < adjI.length; j++) visited[adjI[j]] = 0
  }
}

export { enumerateTriangles }

if (require.main === module) {
  // 1761. 一个图中连通三元组的最小度数
  // https://leetcode.cn/problems/minimum-degree-of-a-connected-trio-in-a-graph/submissions/
  function minTrioDegree(n: number, edges: number[][]): number {
    const deg = new Uint16Array(n)
    for (let i = 0; i < edges.length; i++) {
      edges[i][0]--
      edges[i][1]--
      deg[edges[i][0]]++
      deg[edges[i][1]]++
    }

    let res = 2e15
    enumerateTriangles(n, edges, (a, b, c) => {
      res = Math.min(res, deg[a] + deg[b] + deg[c] - 6)
    })
    return res === 2e15 ? -1 : res
  }

  // n = 6, edges = [[1,2],[1,3],[3,2],[4,1],[5,2],[3,6]]
  console.log(
    minTrioDegree(6, [
      [1, 2],
      [1, 3],
      [3, 2],
      [4, 1],
      [5, 2],
      [3, 6]
    ])
  )
}

File: 7_graph/EnumerateTriangles.go
// EnumeratingC3

// 求出无向图中的所有三元环(a,b,c) a<b<c
//  O(E^1.5)
// https://kopricky.github.io/code/Graph/EnumeratingC3.html

package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

// 枚举所有的三元环,传递回调函数减少空间开销
func enumerateC3(n int, edges [][]int, cb func(u, v, w int)) {
	threshold := 0
	adjList := make([][]int, n)
	for _, e := range edges {
		u, v := e[0], e[1]
		// 无向边定向
		if u < v {
			adjList[u] = append(adjList[u], v)
		} else {
			adjList[v] = append(adjList[v], u)
		}
		threshold++
	}

	threshold = int(math.Sqrt(float64(threshold / 4)))
	memo := make([][]int, n)
	visited := make([]bool, n)

	processHighDegree := func() {
		for i := 0; i < n; i++ {
			if len(adjList[i]) <= threshold {
				continue
			}
			for _, u := range adjList[i] {
				visited[u] = true
			}
			for _, u := range adjList[i] {
				for _, v := range adjList[u] {
					if visited[v] {
						cb(i, u, v)
					}
				}
			}
			for _, u := range adjList[i] {
				visited[u] = false
			}
		}
	}

	processLowDegree := func() {
		for i := 0; i < n; i++ {
			if len(adjList[i]) > threshold {
				continue
			}
			for _, u := range adjList[i] {
				for _, v := range adjList[i] {
					if v > u {
						memo[u] = append(memo[u], i*(1<<32)+v)
					}
				}
			}
		}
		for i := 0; i < n; i++ {
			for _, u := range adjList[i] {
				visited[u] = true
			}
			for j := 0; j < len(memo[i]); j++ {
				hash := memo[i][j]
				a, b := hash>>32, hash&0xffffffff
				if visited[b] {
					cb(a, i, b)
				}
			}
			for _, u := range adjList[i] {
				visited[u] = false
			}
		}
	}

	processHighDegree()
	processLowDegree()
}

func main() {

	// https://judge.yosupo.jp/problem/enumerate_triangles

	const MOD int = 998244353

	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, m int
	fmt.Fscan(in, &n, &m)
	values := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &values[i])
	}

	edges := make([][]int, m)
	for i := 0; i < m; i++ {
		edges[i] = make([]int, 2)
		fmt.Fscan(in, &edges[i][0], &edges[i][1])
	}

	res := 0
	enumerateC3(n, edges, func(u, v, w int) {
		res += values[u] * values[v] % MOD * values[w] % MOD
		res %= MOD
	})
	fmt.Println(res)
}
